query/regexp.go

```diff
@@ -42,3 +43,51 @@ func LowerRegexp(r *syntax.Regexp) *syntax.Regexp {
 
 	return &newRE
 }
+
+// Convert Regexp with captures to Regexp with non captures.
+// Makes a copy leaving orginal Regexp unchanged
+func ConvertCapture(re *syntax.Regexp) (*syntax.Regexp, error) {
```

Comment:
So this is exported because you intend to call `ConvertCapture` from other places in Sourcegraph we create a regex for zoekt?

Alternatively can we call this `OptimizeRegexp`, since there may be more opportunities for us to do rewrites on the query. Additionally, you don't export `hasCapture`, so other packages won't be able to gate behaviour. Rather than two functions, I'd just adjust the implementation of `OptimizeRegexp` to decide if it does anything. Additionally I'd make it not return an error, but instead internally log. This will simplify all the call sites for this, since none of the call sites I imagine will want to fail, but instead fallback to the original regex.

Side-note: The fact that we need to call this is a bit error prone, but I know you tried incorporating this into simplify and it was tricky. So lets move forward with this and see if we can somehow incorporate it into our RPC layers in another PR.

Reply:
+1 for incorporating hasCapture into ConvertCapture. There is no need to export anything in this PR at least. Let's export once we need it?